File: tps/tp01_2026/VerificacaoAnagramaIterativo/VerificacaoAnagramaIterativo.java
public class VerificacaoAnagramaIterativo {

    static boolean saoAnagramas(String s1, String s2) {
        if (s1.length() != s2.length()) return false;

        int[] contador = new int[65536];

        for (int i = 0; i < s1.length(); i++) {
            char c = s1.charAt(i);
            if (c >= 'A' && c <= 'Z') c = (char)(c + 32);
            contador[c]++;
        }

        for (int i = 0; i < s2.length(); i++) {
            char c = s2.charAt(i);
            if (c >= 'A' && c <= 'Z') c = (char)(c + 32);
            contador[c]--;
        }

        for (int i = 0; i < contador.length; i++) {
            if (contador[i] != 0) return false;
        }

        return true;
    }

    static boolean ehFim(String s) {
        return s.equals("FIM");
    }

    public static void main(String[] args) {
        String linha = MyIO.readLine();

        while (!ehFim(linha)) {
            int espaco = linha.indexOf(' ');
            if (espaco != -1) {
                String s1 = linha.substring(0, espaco);
                String s2 = linha.substring(espaco + 1);
                MyIO.println(saoAnagramas(s1, s2) ? "SIM" : "NAO");
            }

            linha = MyIO.readLine();
        }
    }
}

File: tps/tp01_2026/VerificacaoAnagramaIterativo/VerificacaoAnagramaIterativo.c
#include <stdio.h>
#include <string.h>

int saoAnagramas(char *s1, char *s2) {
    int n1 = strlen(s1), n2 = strlen(s2);
    if (n1 != n2) return 0;
    int contador[256] = {0};
    for (int i = 0; i < n1; i++) {
        unsigned char c = s1[i];
        if (c >= 'A' && c <= 'Z') c += 32;
        contador[c]++;
    }
    for (int i = 0; i < n2; i++) {
        unsigned char c = s2[i];
        if (c >= 'A' && c <= 'Z') c += 32;
        contador[c]--;
    }
    for (int i = 0; i < 256; i++) {
        if (contador[i] != 0) return 0;
    }
    return 1;
}

int main() {
    char linha[10000];
    while (fgets(linha, sizeof(linha), stdin)) {
        int len = strlen(linha);
        if (len > 0 && linha[len - 1] == '\n') linha[--len] = '\0';
        if (strcmp(linha, "FIM") == 0) break;
        char s1[5000], s2[5000];
        if (sscanf(linha, "%s %s", s1, s2) == 2) {
            printf(saoAnagramas(s1, s2) ? "SIM\n" : "NAO\n");
        }
    }
    return 0;
}
